src/hss_cache_processor.cpp

```diff
@@ -0,0 +1,258 @@
+ * the license outlined in that COPYING file applies to your use.
+ * Otherwise no rights are granted except for those provided to you by
+ * Metaswitch Networks in a separate written agreement.
+ */
+
+#include "hss_cache_processor.h"
+
+
+HssCacheProcessor::HssCacheProcessor(HssCache* cache) :
+  _cache(cache),
+  _thread_pool(NULL)
+{
+}
+
+bool HssCacheProcessor::start_threads(int num_threads,
+                                      ExceptionHandler* exception_handler,
+                                      unsigned int max_queue)
+{
+  TRC_INFO("Starting threadpool with %d threads", num_threads);
+  _thread_pool = new FunctorThreadPool(num_threads,
+                                       exception_handler,
+                                       exception_callback,
+                                       max_queue);
+
+  return _thread_pool->start();
+}
+
+void HssCacheProcessor::stop()
+{
+  TRC_STATUS("Stopping threadpool");
+  if (_thread_pool)
+  {
+    _thread_pool->stop();
+  }
+}
+
+void HssCacheProcessor::wait_stopped()
+{
+  TRC_STATUS("Waiting for threadpool to stop");
+  if (_thread_pool)
+  {
+    _thread_pool->join();
+    delete _thread_pool; _thread_pool = NULL;
+  }
+}
+
+ImplicitRegistrationSet* HssCacheProcessor::create_implicit_registration_set()
+{
+  return _cache->create_implicit_registration_set();
+}
+
+void HssCacheProcessor::get_implicit_registration_set_for_impu(irs_success_callback success_cb,
+                                                               failure_callback failure_cb,
+                                                               std::string impu,
+                                                               SAS::TrailId trail)
+{
+  // Create a work item that can run on the thread pool, capturing required
+  // variables to complete the work
+  std::function<void()> work = [this, impu, trail, success_cb, failure_cb]()->void {
+    ImplicitRegistrationSet* result = NULL;
+    Store::Status rc = _cache->get_implicit_registration_set_for_impu(impu,
+                                                                      trail,
+                                                                      result);
+
+    if (rc == Store::Status::OK)
+    {
+      success_cb(result);
+    }
+    else
+    {
+      failure_cb(rc);
+    }
+  };
+
+  // Add the work to the pool
+  _thread_pool->add_work(work);
+}
+
+void HssCacheProcessor::get_implicit_registration_sets_for_impis(irs_vector_success_callback success_cb,
+                                                                 failure_callback failure_cb,
+                                                                 std::vector<std::string> impis,
+                                                                 SAS::TrailId trail)
+{
+  // Create a work item that can run on the thread pool, capturing required
+  // variables to complete the work
+  std::function<void()> work = [this, impis, trail, success_cb, failure_cb]()->void {
```

Comment:
Yeah, but they're capturing different variables etc., so I feel like any attempt to commonalize these would just result in them being completely unreadable. This way at least it's clear what they're doing